### PR TITLE
fix: Persist Register/Sign in form states on tab switch [VAN-618]

### PR DESCRIPTION
**Pull request by syedsajjadkazmii**
**Before:**

When the user switched tab between the Register page, Login page, and ForgotPassword page, the entered data in the forms was lost

**After:**

Data entered in any of the forms is persisted on the switching tab between the Register page, Login page, and ForgotPassword page.

[VAN-618](https://2u-internal.atlassian.net/browse/VAN-618)

**Demonstration**

https://user-images.githubusercontent.com/52817156/170024175-ce8bc1ab-5eba-46c7-9f90-208b20c2a0bb.mov

### CODE REVIEW

File: src/register/RegistrationPage.jsx [hunk 1]
```diff
@@ -845,6 +871,24 @@ RegistrationPage.defaultProps = {
     secondaryProviders: [],
     pipelineUserDetails: null,
   },
+  registrationData: {
```
Comment (mubbsharanwar) [inline src/register/RegistrationPage.jsx:874]: It should be `registrationFormData`

File: src/register/tests/RegistrationPage.test.jsx [hunk 2]
```diff
@@ -774,6 +798,115 @@ describe('RegistrationPage', () => {
...
+      store = mockStore({
+        ...initialState,
+        register: {
+          ...initialState.register,
+        },
+      });
+      store.dispatch = jest.fn(store.dispatch);
+      const registrationPage = mount(reduxWrapper(<IntlRegistrationPage {...props} />));
+      registrationPage.find('input#email').simulate('blur', { target: { value: 'test@gmail.con', name: 'email' } });
+      registrationPage.find('RegistrationPage').setState({ emailErrorSuggestion: 'test@gmail.com' });
```
Comment (mubbsharanwar) [inline src/register/tests/RegistrationPage.test.jsx:844]: Why we are setting `emailErrorSuggestion` state itself? Should it not set on onBlure when user enter incorrect email?

**Review by mubbsharanwar — COMMENTED** [review]
There are some suggestions otherwise its looking good.

File: src/forgot-password/data/actions.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import { AsyncActionType } from '../../data/utils';
 
 export const FORGOT_PASSWORD = new AsyncActionType('FORGOT', 'PASSWORD');
+export const FORGOT_PASSWORD_SET_FORM_DATA = 'FORGOT_PASSWORD_PERSIST_FORM_DATA';
```
[the remark this answers is not in the document]
Comment (syedsajjadkazmii) [reply src/forgot-password/data/actions.js:4]: Actually `AsyncActionType` is being used for the action types of asynchronous methods, where we have a BEGIN, SUCCESS, FAILURE states, but the `FORGOT_PASSWORD_PERSIST_FORM_DATA` is not asynchronous method, its simply setting data in store.
(Based on the action type used for `REGISTER_CLEAR_USERNAME_SUGGESTIONS` used in register page.)

File: src/login/data/actions.js [hunk 4]
```diff
@@ -25,3 +26,9 @@ export const loginRequestFailure = (loginError) => ({
 export const loginRequestReset = () => ({
   type: LOGIN_REQUEST.RESET,
 });
+
+// Persist Form Data
```
Comment (mubbsharanwar) [inline src/login/data/actions.js:30]: nit: remove this comment.

File: src/register/RegistrationPage.jsx [hunk 5]
```diff
@@ -16,7 +16,7 @@ import {
 import { Error, Close } from '@edx/paragon/icons';
 import FormFieldRenderer from '../field-renderer';
 import {
-  clearUsernameSuggestions, registerNewUser, resetRegistrationForm, fetchRealtimeValidations,
+  clearUsernameSuggestions, registerNewUser, resetRegistrationForm, fetchRealtimeValidations, setRegisterFormData,
```
Comment (mubbsharanwar) [inline src/register/RegistrationPage.jsx:19]: should it not be `setRegistrationFormData`?

File: src/register/RegistrationPage.jsx [hunk 6]
```diff
@@ -26,6 +26,7 @@ import {
   registrationRequestSelector,
   validationsSelector,
   usernameSuggestionsSelector,
+  registerFormDataSelector,
```
Comment (mubbsharanwar) [inline src/register/RegistrationPage.jsx:29]: nit:
`registerFormDataSelector` to `registrationFormDataSelector`

File: src/register/data/actions.js [hunk 7]
```diff
@@ -53,3 +54,9 @@ export const fetchRealtimeValidationsFailure = () => ({
 export const clearUsernameSuggestions = () => ({
   type: REGISTER_CLEAR_USERNAME_SUGGESTIONS,
 });
+
+// Persist Form Data
```
Comment (mubbsharanwar) [inline src/register/data/actions.js:58]: nit: remove this comment

**Review by mubbsharanwar — APPROVED** [review]
Some suggestions otherwise looks good.

**Comment by waheedahmed** [comment]
It broke the register form on production, reverted.

**Pull request merged**